sys/include/net/gnrc/lorawan.h

```diff
@@ -0,0 +1,251 @@
+
+#define GNRC_LORAWAN_REQ_STATUS_SUCCESS (0)     /**< MLME or MCPS request successful status */
+#define GNRC_LORAWAN_REQ_STATUS_DEFERRED (1)    /**< the MLME or MCPS confirm message is asynchronous */
+
+/**
+ * @brief MAC Information Base attributes
+ */
+typedef enum {
+    MIB_ACTIVATION_METHOD
+} mlme_mib_type_t;
+
+/**
+ * @brief MAC Information Base descriptor for MLME Request-Confirm
+ */
+typedef struct {
+    mlme_mib_type_t type; /**< MIB attribute identifier */
+    union {
+        int activation;
+    } param;              /**< parameters of the MIB get-set request */
+} mlme_mib_t;
+
+/**
+ * @brief MAC (sub) Layer Management Entity (MLME) request representation
+ */
+typedef struct {
+    union {
+        mlme_lorawan_join_t join;
+        mlme_mib_t mib;
+    } param;        /**< parameters of the MLME request */
+    uint8_t type;   /**< type of the MLME request */
+} mlme_request_t;
+
+/**
+ * @brief Mac Common Part Sublayer (MCPS) request representation
+ */
+typedef struct {
+    union {
+        mcps_data_t data;
+    } param;        /**< parameters of the MCPS request */
+    uint8_t type;   /**< type of the MCPS request */
+} mcps_request_t;
+
+/**
+ * @brief MAC (sub) Layer Management Entity (MLME) confirm representation
+ */
+typedef struct {
+    int16_t status; /**< status of the MLME confirm */
+    uint8_t type;   /**< type of the MLME confirm */
+    union {
+        mlme_link_req_confirm_t link_req;
+        mlme_mib_t mib;
+    } param; /**< parameters of the MLME confirm */
+} mlme_confirm_t;
+
+/**
+ * @brief Mac Common Part Sublayer (MCPS) confirm representation
+ */
+typedef struct {
+    void *data;     /**< data of the MCPS confirm */
+    int16_t status; /**< status of the MCPS confirm */
+    uint8_t type;   /**< type of the MCPS confirm */
+} mcps_confirm_t;
+
+/**
+ * @brief Mac Common Part Sublayer (MCPS) indication representation
+ */
+typedef struct {
+    uint8_t type; /**< type of the MCPS indication */
+    union {
+        mcps_data_t data;
+    } param; /**< params of the MCPS indication */
+} mcps_indication_t;
+
+/**
+ * @brief MAC (sub) Layer Management Entity (MLME) indication representation
+ */
+typedef struct {
+    uint8_t type; /**< type of the MLME indication */
+} mlme_indication_t;
+
+/**
+ * @brief Indicate the MAC layer there was a timeout event
+ *
+ * @param[in] mac pointer to the MAC descriptor
+ */
+void gnrc_lorawan_event_timeout(gnrc_lorawan_t *mac);
+
+/**
+ * @brief Indicate the MAC layer when the transmission finished
+ *
+ * @param[in] mac pointer to the MAC descriptor
+ */
+void gnrc_lorawan_event_tx_complete(gnrc_lorawan_t *mac);
+
+/**
+ * @brief Init GNRC LoRaWAN
+ *
+ * @param[in] mac pointer to the MAC descriptor
+ * @param[in] nwkskey buffer to store the NwkSKey. Should be at least 16 bytes long
+ * @param[in] appskey buffer to store the AppsKey. Should be at least 16 bytes long
+ */
+void gnrc_lorawan_init(gnrc_lorawan_t *mac, uint8_t *nwkskey, uint8_t *appskey);
```

Comment:
can (some) parameters be made `const`? Check for all functions please.

Reply:
yes, most of them. I will check

Reply:
done! Please mark as reviewed if OK